src/note_position.rs

```diff
@@ -0,0 +1,92 @@
+use std::fmt;
+use std::iter::Peekable;
+
+/// An iterator adaptor that wraps each element in an [`Position`](../enum.Position.html).
+///
+/// Iterator element type is `Position<I::Item>`.
+/// This iterator is *fused*.
+///
+/// See [`.note_position()`](../trait.Itertools.html#method.note_position) for more information.
+pub struct NotePosition<I>
+    where I: Iterator,
+{
+    handled_first: bool,
+    peekable: Peekable<I>,
+}
+
+/// Create a new `NotePosition` iterator.
+pub fn note_position<I>(iter: I) -> NotePosition<I>
+    where I: Iterator,
+{
+    NotePosition {
+        handled_first: false,
+        peekable: iter.peekable(),
+    }
+}
+
+/// A value yielded by `NotePosition`.
+/// Indicates the position of this element in the iterator results.
+///
+/// See [`.note_position()`](trait.Itertools.html#method.note_position) for more information.
+pub enum Position<T> {
+    /// This is the first element.
+    First(T),
+    /// This is neither the first nor the last element.
+    Middle(T),
+    /// This is the last element.
+    Last(T),
+    /// This is the only element.
+    Only(T),
+}
+
+impl<T: PartialEq> PartialEq for Position<T> {
+    fn eq(&self, other: &Position<T>) -> bool {
+        match (self, other) {
+            (&Position::First(ref a), &Position::First(ref b)) => a == b,
+            (&Position::Middle(ref a), &Position::Middle(ref b)) => a == b,
+            (&Position::Last(ref a), &Position::Last(ref b)) => a == b,
+            (&Position::Only(ref a), &Position::Only(ref b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Position<T> {
```

Comment:
This is equivalent to derive(Debug) except for pretty printing, why not use deriving.